### PR TITLE
added template diff tracking and history

### PR DESCRIPTION
**Pull request by imanassypov**
Added functionality:
- per-template file, add comment header with last diff
- per-template file, summary page - last 'n' commit messages history

Notes:
- recognize comment char for J2 vs Velocity
- truncate summary section to match max supported length in dnac of 1024
- variable defined DNAC_URL
- added runner build stage

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -20,6 +37,8 @@ validate:
   stage: validate
   script:
     - python scripts/validate.py
+  tags:
+    - macos
```
Comment (oboehmer) [inline .gitlab-ci.yml:43]: what do you use these tags for?

File: deployment/Management-Access.yaml [hunk 2]
```diff
@@ -4,7 +4,4 @@
 
 
 devices:
-  CSR-1:
-  CSR-2:
-  CSR-3:
-  CSR-4:
+  wlc01:
```
Comment (oboehmer) [inline deployment/Management-Access.yaml]: I would rather keep those devices in the repository, can you please revert the changes? Maybe you can apply them in your fork?

File: scripts/DNACTemplate.py [hunk 3]
```diff
@@ -58,6 +69,54 @@ def __init__(self, config_file=None, project=None, connect=True):
...
+        '''
+        template_comments_jinja = "{{## {} ##}}\n"
+        template_comments_velocity = '## {}\n'
+        template_comments = template_comments_jinja if (language=='JINJA') else template_comments_velocity
+
+        file_diff = ""
+        diff_comments = ""
+
+        try:
+            repo = Git(repo_path)
```
Comment (oboehmer) [inline scripts/DNACTemplate.py]: maybe do this in `__init__` so you avoid doing this multiple times?

File: scripts/DNACTemplate.py [hunk 4]
```diff
@@ -58,6 +69,54 @@ def __init__(self, config_file=None, project=None, connect=True):
...
+            file_diff = repo.log('--pretty=%H','-p','-1', filename)
+            logger.info('git log for file {} loaded at {} chars'.format(repo_path,
+                                                                        len(file_diff)))
+        except exc.GitError as e:
+            logger.fatal('Could not load repository at {} file {}:'.format(repo_path,
+                                                                           filename))
```
Comment (oboehmer) [inline scripts/DNACTemplate.py]: do we want to fail the script execution if we can't do the comments/updates?

File: scripts/DNACTemplate.py [hunk 5]
```diff
@@ -210,12 +269,22 @@ def provision_templates(self, template_dir, purge=True, result_json=None):
...
+                                                 commits_count=int(self.config.commit_history_count))
+
+                template_diff = self.get_file_diff(repo_path=self.config.git_root,
+                                                   filename=os.path.join(template_dir, template_file),
+                                                   language = language)
                 # DNAC requires includes to include the absolute path, so we make this
                 # dependent on the project (i.e. {% include "__PROJECT__/foo" %} )
-                template_content = re.sub('__PROJECT__', self.template_project, template_content)
+                template_content = template_diff + \
+                                   re.sub('__PROJECT__', self.template_project, template_content)
```
Comment (oboehmer) [inline scripts/DNACTemplate.py]: maybe better
```suggestion
                template_content = '{}{}'.format(template_diff,
                                   re.sub('__PROJECT__', self.template_project, template_content))
```

File: scripts/DNACTemplate.py [hunk 6]
```diff
@@ -280,10 +351,15 @@ def provision_templates(self, template_dir, purge=True, result_json=None):
             if not template_id:
                 raise Exception('Creation of template "{}" failed: {}'.format(template_name, data))
 
+            # Template comments length is limited to COMMENTS_MAX_CHARS
+            comments = (('committed by gitlab-ci at {} UTC\n').format(datetime.utcnow()) + \
+                         commit_log)
+            comments = (comments[:self.COMMENTS_MAX_CHARS]) if len(comments) > self.COMMENTS_MAX_CHARS else comments
```
Comment (oboehmer) [inline scripts/DNACTemplate.py]: should be enough? 
```suggestion
            comments = comments[:self.COMMENTS_MAX_CHARS]
```

File: scripts/DNACTemplate.py [hunk 7]
```diff
@@ -36,12 +37,22 @@ def _basename(path):
 
 class DNACTemplate(object):
 
+    #fixed CHAR length of comments field in DNAC Template
+    COMMENTS_MAX_CHARS = 1024
```
Comment (oboehmer) [inline scripts/DNACTemplate.py]: should we make this configurable (i.e. fetch it from config file?)

**Review by oboehmer — CHANGES_REQUESTED** [review]
sorry for the delay, started with the review and forgot about it ;-)  a few comments and requests..

**Comment by imanassypov** [comment]
Oliver - thank you for great feedback, made changes as requested

**Review by oboehmer — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**